pkg/k8shandler/statefulset.go

```diff
@@ -2,98 +2,387 @@ package k8shandler
+	apps "k8s.io/api/apps/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	"k8s.io/apimachinery/pkg/runtime"
 )
 
 type statefulSetNode struct {
-	resource apps.StatefulSet
+	self apps.StatefulSet
+	// prior hash for configmap content
+	configmapHash string
+	// prior hash for secret content
+	secretHash string
+
+	clusterName       string
+	clusterSize       int32
+	priorReplicaCount int32
 }
 
-func (node *statefulSetNode) getResource() runtime.Object {
-	return &node.resource
+func (statefulSetNode *statefulSetNode) populateReference(nodeName string, node v1alpha1.ElasticsearchNode, cluster *v1alpha1.Elasticsearch, roleMap map[v1alpha1.ElasticsearchNodeRole]bool, replicas int32) {
+
+	labels := newLabels(cluster.Name, roleMap)
+
+	statefulSet := apps.StatefulSet{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "StatefulSet",
+			APIVersion: apps.SchemeGroupVersion.String(),
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      nodeName,
+			Namespace: cluster.Namespace,
+			Labels:    labels,
+		},
+	}
+
+	statefulSet.Spec = apps.StatefulSetSpec{
+		Replicas: &replicas,
+		Selector: &metav1.LabelSelector{
+			MatchLabels: newLabelSelector(cluster.Name, roleMap),
+		},
+		Template: newPodTemplateSpec(nodeName, cluster.Name, cluster.Namespace, node, cluster.Spec.Spec, labels, roleMap),
+	}
+	statefulSet.Spec.Template.Spec.Containers[0].ReadinessProbe = nil
+
+	addOwnerRefToObject(&statefulSet, getOwnerRef(cluster))
+
+	statefulSetNode.self = statefulSet
+	statefulSetNode.clusterName = cluster.Name
 }
 
-func (node *statefulSetNode) getRevision(cfg *desiredNodeState) (string, error) {
-	return "", nil
+func (node *statefulSetNode) getNodeState() v1alpha1.ElasticsearchNodeStatus {
+	rolloutForReload := v1alpha1.ConditionFalse
+	rolloutForUpdate := v1alpha1.ConditionFalse
+
+	// see if we need to update the deployment object
+	if node.isChanged() {
+		rolloutForUpdate = v1alpha1.ConditionTrue
+	}
+
+	// check if the configmapHash changed
+	/*newConfigmapHash := getConfigmapDataHash(node.clusterName, node.self.Namespace)
+	if newConfigmapHash != node.configmapHash {
+		rolloutForReload = v1alpha1.ConditionTrue
+	}*/
+
+	// check if the secretHash changed
+	newSecretHash := getSecretDataHash(node.clusterName, node.self.Namespace)
+	if newSecretHash != node.secretHash {
+		rolloutForReload = v1alpha1.ConditionTrue
+	}
+
+	return v1alpha1.ElasticsearchNodeStatus{
+		StatefulSetName: node.self.Name,
+		UpgradeStatus: v1alpha1.ElasticsearchNodeUpgradeStatus{
+			ScheduledForUpgrade:  rolloutForUpdate,
+			ScheduledForRedeploy: rolloutForReload,
+		},
+	}
 }
 
-func (node *statefulSetNode) awaitingRollout(cfg *desiredNodeState, currentRevision string) (bool, error) {
-	return false, nil
+func (node *statefulSetNode) getName() string {
+	return node.self.Name
 }
 
-func (node *statefulSetNode) needsPause(cfg *desiredNodeState) (bool, error) {
-	return false, nil
+func (node *statefulSetNode) waitForNodeRejoinCluster() (error, bool) {
```

Comment:
Is this really different from the same method for other node types?

Reply:
no, but since we're using an interface it lets us do `node.waitForNodeRejoinCluster` without needing to pull the object out and pass in a bunch of variables.

Reply:
That was the point I was trying to get at is that seems like it should be possible to have `watiForNodeRejoinCluster` that does the same flow regardless of the actual node type without more or less re-implementing it;  that's the whole point of interfaces.

Reply:
So, if i try moving `waitForNodeRejoinCluster` up to the `nodetypefactory` level then at the `deploymentNode` and `statefulsetNode` level they are unable to find that function for some reason as well as someother errors... [1]
```
_output/src/github.com/openshift/elasticsearch-operator/pkg/k8shandler/deployment.go:303:19: node.waitForNodeLeaveCluster undefined (type *deploymentNode has no field or method waitForNodeLeaveCluster)
_output/src/github.com/openshift/elasticsearch-operator/pkg/k8shandler/deployment.go:318:20: node.waitForNodeRejoinCluster undefined (type *deploymentNode has no field or method waitForNodeRejoinCluster)
_output/src/github.com/openshift/elasticsearch-operator/pkg/k8shandler/deployment.go:420:20: node.waitForNodeRejoinCluster undefined (type *deploymentNode has no field or method waitForNodeRejoinCluster)
_output/src/github.com/openshift/elasticsearch-operator/pkg/k8shandler/nodetypefactory.go:101:6: invalid receiver type NodeTypeInterface (NodeTypeInterface is an interface type)
```
[1] https://github.com/openshift/elasticsearch-operator/pull/80/commits/9b6108a525dc0c1bf1f20fb2d17ba75b0abb2c8c

Reply:
So you could either define the method on `NodeTypeInterface` or a separate Interface for these specific types. Recall that golang only requires the the thing impl the method.  Maybe something to further investigate after this merge